### PR TITLE
improve order map w/ mapbox & zoom capabilities; closes #43

### LINKED ISSUE #43: Improvements to Map
This isn't a priority, but we discussed a few (future) improvements to make to the map:

  - we should be able to filter by the order type
  - I would (personally) like to see a design that allows for zoom. For example, I made this (sort of crappy) [body map](http://vsoch.github.io/bodymap/map.html) in grad school to visualize a particularly morbid dataset - something like this would do well for the orders.

### PR DESCRIPTION
**Pull request by shea256**
Addresses #43

### CODE REVIEW

**Comment by shea256** [comment]
Here's a screenshot of what the new map looks like:

Screen Shot 2019-09-25 at 3 02 43 AM

File: fg/apps/main/templates/maps/orders.html [hunk 1]
```diff
@@ -1,64 +1,74 @@
...
-{% endblock %}
-{% block pagescripts %}
-<script src="//cdnjs.cloudflare.com/ajax/libs/d3/3.5.3/d3.min.js"></script>
-<script src="//cdnjs.cloudflare.com/ajax/libs/topojson/1.6.9/topojson.min.js"></script>
-<script src="//cdnjs.cloudflare.com/ajax/libs/datamaps/0.5.9/datamaps.world.min.js"></script>
+<script src='https://api.tiles.mapbox.com/mapbox-gl-js/v1.3.1/mapbox-gl.js'></script>
+<link href='https://api.tiles.mapbox.com/mapbox-gl-js/v1.3.1/mapbox-gl.css' rel='stylesheet' />
 <script>
+  mapboxgl.accessToken = 'pk.eyJ1Ijoic2hlYTI1NiIsImEiOiJjazB5dXJqYTgwN3p5M2NvODhrYWY4ejBuIn0.WYrEJAAKQBR3i38SU3nllQ';
+
```
Comment (vsoch) [inline fg/apps/main/templates/maps/orders.html]: This looks great! Before I test, could you please:

 - confirm the [scopes](https://docs.mapbox.com/help/how-mapbox-works/access-tokens/) of the token that you've created
 - have the token provided by the user in the settings/config.py and add to the view - we would never want to require the user to edit the view (html) to update it.
 - add documentation to the map docs to tell the user about how the map is generated, link to the mapbox, and instruct how to generate, scope, and update the token (the one provided in the config.py settings should be faux (e.g.) `MAPBOX_TOKEN=xxxxxxxxxxxx`)
 - If you are adding metadata about an order, (right now all are hard coded as "OpenEnzyme" and a dummy description) take a look at how we use the Shippo API to obtain it, and make a suggestion (even if you can't test) for how we can update the data extraction function to obtain what is needed. If you update this (fg/apps/main/management/commands/generate_mapdata.py) I can test it for you on the server, and provide an updated ordercoords.json to test from.

 - update the screen shot in the docs to be the current map
This will be a general pattern that we want to follow - if something is added that warrants a token (or the user to understand something) we typically want to add it to the docs (in the docs/_docs) folder. There is already a section for the map, so you should be able to update it.

I'll be able to preview/test this out when we get some of the current updates finished up. Thanks for working on this!

Comment (vsoch) [reply fg/apps/main/templates/maps/orders.html]: oh and this is hugely important - (and you actually reminded me in your issue comment!) We need a map that supports markers that can be scaled to the number of orders (akin to the current) or a heatmap style that does similar. What we can't do is have individual markers for each order - that kind of map is more appropriate for locating single entities, but not representing size/quantity. I don't make a ton of maps, but if you look at the  [code here](http://vsoch.github.io/bodymap/) when you click map it supports zooming too, and if I remember correctly you can use that map style for more than just individual points.

Comment (vsoch) [reply fg/apps/main/templates/maps/orders.html]: I was going to start with that to refactor the map (and address this issue) - of course not styling so dark :P That particular dataset I found randomly and is fairly dark, I just made the map with the body for fun. But it's dark.

Comment (vsoch) [reply fg/apps/main/templates/maps/orders.html]: And to answer your question - yes we want to eventually have filters based on order types, so keep that in mind as you design. Okay back to work!

Comment (shea256) [reply fg/apps/main/templates/maps/orders.html]: Hey thanks @vsoch! Please excuse the delay. Please find my responses inline below:

> confirm the scopes of the token that you've created

This is actually the default public token so it is the same for everyone. I will document this.

> have the token provided by the user in the settings/config.py and add to the view - we would never want to require the user to edit the view (html) to update it.

Will do. I will set the default token for the user, and if they want to change it to a registered token they can.

> add documentation to the map docs to tell the user about how the map is generated, link to the mapbox, and instruct how to generate, scope, and update the token (the one provided in the config.py settings should be faux (e.g.) MAPBOX_TOKEN=xxxxxxxxxxxx)

Will do, thanks. Didn't want to do this until getting some level of feedback from you, just in case I had to go down a different route.

> If you are adding metadata about an order, (right now all are hard coded as "OpenEnzyme" and a dummy description) take a look at how we use the Shippo API to obtain it, and make a suggestion (even if you can't test) for how we can update the data extraction function to obtain what is needed. If you update this (fg/apps/main/management/commands/generate_mapdata.py) I can test it for you on the server, and provide an updated ordercoords.json to test from.

OK this is exactly what I wanted to do, but was unable to run that command. It would be fairly hard to accomplish this if I have to rely on you to generate the `ordercoords.json` file. Is there documentation on how I can run it? If so, I can handle the additional data tagging.

> update the screen shot in the docs to be the current map
This will be a general pattern that we want to follow - if something is added that warrants a token (or the user to understand something) we typically want to add it to the docs (in the docs/_docs) folder. There is already a section for the map, so you should be able to update it.

Sure thing!

> oh and this is hugely important - (and you actually reminded me in your issue comment!) We need a map that supports markers that can be scaled to the number of orders (akin to the current) or a heatmap style that does similar. What we can't do is have individual markers for each order - that kind of map is more appropriate for locating single entities, but not representing size/quantity. 

Sounds good.

> And to answer your question - yes we want to eventually have filters based on order types, so keep that in mind as you design. Okay back to work!

Makes sense. Will keep that in mind so that we can add an extension to the design in the future.

Thanks!

Comment (vsoch) [reply fg/apps/main/templates/maps/orders.html]: > This is actually the default public token so it is the same for everyone. I will document this.

Oh that's really interesting! I've never seen a shared API token - I wonder how they track usage per client? Yes, please add to the documentation, and include limits (I'm guessing there are some) and how to update the token (in the settings/config.py) if the user wants to get a registered token.

> OK this is exactly what I wanted to do, but was unable to run that command. It would be fairly hard to accomplish this if I have to rely on you to generate the ordercoords.json file. Is there documentation on how I can run it? If so, I can handle the additional data tagging.

You would need a Shippo API token - and actually not a test token, but a production token, and I don't think I can give this to you, but I can get the data you need from the API for you. Here is what I just ran for you (a subset of the file, manually):

```python
from fg.apps.main.utils import load_json
from fg.settings import SHIPPO_TOKEN

import os
import json
import requests
import shutil
import sys

url="https://api.goshippo.com/orders/"
headers={"Authorization": "ShippoToken %s" % SHIPPO_TOKEN}

response = requests.get(url, headers=headers)
listing = []

while response.status_code == 200:

            results = response.json()
            if "results" not in results:
                print("No results in response.")
                sys.exit(1)

            listing = listing + results['results']
            response = requests.get(results['next'], headers=headers)
```
I've dumped to json, and I can send this to you through your private email with names removed (please send this to me).

This script is typically run with `python manage.py generate_mapdata` but I would recommend loading the data that I send you and doing testing manually. [Here (https://github.com/vsoch/freegenes/blob/master/fg/apps/main/management/commands/generate_mapdata.py) is a direct link to the script and it's run via [this cron job](https://github.com/vsoch/freegenes/blob/master/Dockerfile#L58) that is created on the container build. Per your request, I've updated the [orders map](https://vsoch.github.io/freegenes/docs/usage/interface#orders-map) with this complete information.

Thanks @shea256 !

**Comment by shea256** [comment]
To-do list for this PR:

- [x] have the token provided by the user in the settings/config.py and add to the view - we would never want to require the user to edit the view (html) to update it.
- [x] We need a map that supports markers that can be scaled to the number of orders (akin to the current) or a heatmap style that does similar.
- [ ] add documentation to the map docs to tell the user about how the map is generated, link to the mapbox, and instruct how to generate, scope, and update the token (the one provided in the config.py settings should be faux (e.g.) MAPBOX_TOKEN=xxxxxxxxxxxx)
- [ ] If you are adding metadata about an order, (right now all are hard coded as "OpenEnzyme" and a dummy description) take a look at how we use the Shippo API to obtain it, and make a suggestion (even if you can't test) for how we can update the data extraction function to obtain what is needed. If you update this (fg/apps/main/management/commands/generate_mapdata.py) I can test it for you on the server, and provide an updated ordercoords.json to test from.
- [ ] update the screen shot in the docs to be the current map

File: fg/apps/main/templates/maps/orders.html (hunk 1 above)
Comment (shea256) [reply fg/apps/main/templates/maps/orders.html]: @vsoch Submitted a request for review. Let me know when you've gotten a chance to review and whether you give it the green light. Thanks!

**Comment by vsoch** [comment]
hey @shea256 - you first should rebase with master to resolve the conflicts above! Also note that when you request for review, I'm automatically notified:

image

So no need to do it again :)

File: .gitignore [hunk 2]
```diff
@@ -19,6 +19,7 @@ OLD
 secrets.py
 local_settings.py
 cronjob
+.DS_Store
```
Comment (vsoch) [inline .gitignore]: Good call here!

File: docs/_docs/development/setup.md [hunk 3]
```diff
@@ -113,6 +113,22 @@ FREEGENES_TWIST_EUTOKEN=xxxxxxxxxxxxxxxxxxxxxx
 If you don't define these variables, they will be set to None and the
 option to import orders from Twist simply won't be presented.
 
+### Map Tokens
+
+The Orders Map (a map of all historical orders found at the route /map/orders) is generated using Mapbox GL JS (https://docs.mapbox.com/mapbox-gl-js/api/).
+
+In order for the JS library to generate the map, you need to supply a valid Mapbox access token ("mapboxgl.accessToken").
+
```
Comment (vsoch) [inline docs/_docs/development/setup.md]: I would say `mapboxgl.accessToken` in the JavaScript that creates the map so it's clear - we have lots of different references for variable, and it should be clear where / what language you are referring to.

File: fg/settings/config.py [hunk 4]
```diff
@@ -106,3 +106,7 @@
 #    'pam_auth',
 #    'saml_auth'
 ]
+
+# Map
+## Default / public Mapbox token
+MAPBOX_ACCESS_TOKEN = 'pk.eyJ1Ijoic2hlYTI1NiIsImEiOiJjazB5dXJqYTgwN3p5M2NvODhrYWY4ejBuIn0.WYrEJAAKQBR3i38SU3nllQ'
```
Comment (vsoch) [inline fg/settings/config.py]: perfect!

File: docs/_docs/development/setup.md [hunk 5]
```diff
@@ -113,6 +113,22 @@ FREEGENES_TWIST_EUTOKEN=xxxxxxxxxxxxxxxxxxxxxx
...
+This token is passed from the server to the Orders Map template and can be set as a configuration variable.
+
+By default, the default public Mapbox token is used and defined in "/fg/settings/config.py". This will work perfectly fine for all the basic map features, but if you are expanding the feature set of the map, you will need to register a Mapbox token and replace the default public one with your own.
+
+You can set your own token in the "secrets.py" file you created as follows:
+
+```python
+MAPBOX_ACCESS_TOKEN = 'XXXXX'
```
Comment (vsoch) [inline docs/_docs/development/setup.md]: Nicely said here.

**Comment by vsoch** [comment]
It looks like the merge conflict is for docs, so I'm definitely happy to test and review now.

**Comment by shea256** [comment]
> It looks like the merge conflict is for docs, so I'm definitely happy to test and review now.

OK sounds good.

File: fg/apps/main/templates/maps/orders.html [hunk 6]
```diff
@@ -1,64 +1,76 @@
...
+    style: 'mapbox://styles/mapbox/light-v10',
+    center: [-95, 40],
+    zoom: 3,
+  });
+
+  var geojson = {
+    type: 'FeatureCollection',
+    features: [
+    {% for zip,datum in data.items %}
+      {
```
Comment (vsoch) [inline fg/apps/main/templates/maps/orders.html]: Always put the for loop on the same line as the first item in content, otherwise you get a page with a lot of empty spaces. E.g., instead of:

```
{% for thing in things %}
{{ thing }}
{% endfor %}
```
You should do

```
{% for thing in things %}{{ thing }}{% endfor %}
```
In this case, it's just about moving the opening brace up so it's after the for loop start, and then moving the closing brace to follow suit.

File: fg/apps/main/templates/maps/orders.html [hunk 7]
```diff
@@ -1,64 +1,76 @@
...
+        properties: {
+          title: 'OpenEnzyme',
+          description: 'A test shipment of OpenEnzyme to a community lab.',
+        },
+        size: {
+          width: Math.sqrt({{datum.count}})*15,
+          height: Math.sqrt({{datum.count}})*15,
+        }
+      },
+    {% endfor %}
```
Comment (vsoch) [inline fg/apps/main/templates/maps/orders.html]: This endfor should be moved up so it's directly after the comma.

Also - you should be checking for `{% if forloop.last %}` to *not* render the comma.

File: fg/apps/main/templates/maps/orders.html [hunk 8]
```diff
@@ -1,64 +1,76 @@
...
+    ]
+  };
+
+  geojson.features.forEach(function(marker) {
+    var el = document.createElement('div');
+    el.className = 'marker';
+    el.setAttribute('style', 'width: ' + marker.size.width + 'px; height: ' + marker.size.height + 'px;');
+
```
Comment (vsoch) [inline fg/apps/main/templates/maps/orders.html]: Please add a popup to each marker with some trivial information - we can improve upon it if needed. Minimally the user should know the zip code, and number of orders from it.

File: .gitignore (hunk 2 above)
Comment (shea256) [reply .gitignore]: Thanks!

File: docs/_docs/development/setup.md (hunk 3 above)
Comment (shea256) [reply docs/_docs/development/setup.md]: I'm a bit confused. Can you explain what you mean?

File: fg/settings/config.py (hunk 4 above)
Comment (shea256) [reply fg/settings/config.py]: Thanks!

File: docs/_docs/development/setup.md (hunk 5 above)
Comment (shea256) [reply docs/_docs/development/setup.md]: Thank you :)

File: fg/apps/main/templates/maps/orders.html (hunk 6 above)
Comment (shea256) [reply fg/apps/main/templates/maps/orders.html]: Oh ok. I did it that way because the other way didn't seem to be easy to visually parse / understand. But I will follow your conventions / guidance here and update.

File: fg/apps/main/templates/maps/orders.html (hunk 7 above)
Comment (shea256) [reply fg/apps/main/templates/maps/orders.html]: > This endfor should be moved up so it's directly after the comma.

Will make this change as mentioned above.

> Also - you should be checking for {% if forloop.last %} to not render the comma.

I chose not to do this because it overcomplicates the code and because JavaScript objects both allow and encourage trailing commas for simplicity and in order to make it harder to make editing errors. This is in contrast to the JSON specification which prohibits trailing commas.

File: fg/apps/main/templates/maps/orders.html (hunk 8 above)
Comment (shea256) [reply fg/apps/main/templates/maps/orders.html]: Sure thing, will do!

File: docs/_docs/development/setup.md (hunk 3 above)
Comment (vsoch) [reply docs/_docs/development/setup.md]: Write this:

> In order for the JS library to generate the map, you need to supply a valid Mapbox access token ("mapboxgl.accessToken" in the JavaScript that creates the map). This variable is populated from the `MAPBOX_ACCESS_TOKEN` in your `settings/config.py`:

File: fg/apps/main/templates/maps/orders.html (hunk 6 above)
Comment (vsoch) [reply fg/apps/main/templates/maps/orders.html]: If you want to see what I mean, do a "view source" on a page before and after. You can also look at almost any jekyll site and see a huge amount of white space because folks don't tend to put the logic statements on the same line as the code they are wrapping.

File: fg/apps/main/templates/maps/orders.html (hunk 7 above)
Comment (vsoch) [reply fg/apps/main/templates/maps/orders.html]: Oh that's really interesting - I wasn't aware of this! Yes I'm totally okay with that.

**Comment by vsoch** [comment]
hey @shea256 I'm not sure how you are using git, but you just pushed 49 commits (that were already integrated into master) and I want to suggest that you take the changes you've made and create a new pull request, likely manually at this point because I don't know what you've done locally.  We will not be able to merge this with the extraneous commits.

File: fg/apps/main/templates/maps/orders.html (hunk 8 above)
Comment (shea256) [reply fg/apps/main/templates/maps/orders.html]: This is done. Please see below.

File: docs/_docs/development/setup.md (hunk 3 above)
Comment (shea256) [reply docs/_docs/development/setup.md]: This is done.

File: fg/apps/main/templates/maps/orders.html (hunk 6 above)
Comment (shea256) [reply fg/apps/main/templates/maps/orders.html]: This is done.

**Comment by shea256** [comment]
@vsoch All comments addressed and awaiting your approval.

**Comment by shea256** [comment]
Hey @vsoch! All items on the checklist should now be addressed. Please let me know if there’s anything else you need here.

**Comment by shea256** [comment]
Specifically, I just added zoom controls, fixed the popups, changed the marker colors to FreeGenes green, and expanded the map to the full width of the page.

Thanks for the review!

File: fg/settings/logging.py [hunk 9]
```diff
@@ -14,6 +14,11 @@
 ENABLE_SENTRY=False
 SENTRY_DSN="https://xxxxxxxxxxxxxxxxxxxxxxx@sentry.io/xxxxxxx"
 
+try:
+    from .secrets import *
+except ImportError:
+    pass
+
 # Default Django logging is WARNINGS+ to console
```
Comment (vsoch) [inline fg/settings/logging.py]: Why did you add this here? Secrets should be import from config.py (at the end) and the main entrypoint is of course the __init__.py. This should be removed.

File: requirements.txt [hunk 10]
```diff
@@ -42,7 +42,7 @@ retrying
 rq-scheduler
 shapely
 sendgrid==6.1.0
-sentry-sdk==0.11.2
+sentry-sdk==0.12.3
 shippo==2.0.0rc1
```
Comment (vsoch) [inline requirements.txt]: Why did you change the version here?

Comment (shea256) [reply requirements.txt]: I was getting an error that was crashing my worker. It traced back to an issue with the Sentry token. The bug was fixed with an upgrade.

File: fg/settings/logging.py (hunk 9 above)
Comment (shea256) [reply fg/settings/logging.py]: In your documentation you say to copy config.py and create a new file called secrets.py.

The problem with this file is that SENTRY_DSN is set right before later code that uses the variable. Thus it will override any previously set value.

**Comment by vsoch** [comment]
For local testing, `ENABLE_SENTRY` should be False (see logging.py) so you shouldn't need a token, nor should the block that imports the sdk be used. Did you set up logging for your localhost??

**Comment by shea256** [comment]
I did not set it up and the value is indeed set to false. That said, for some reason the code block appears to be executed and the worker crashes. This is the fix that worked. I can remove these lines if you would like and I can develop locally with my slight modification. Additionally I can investigate further to get to the bottom of why that fix worked.

**Comment by vsoch** [comment]
Could you try removing the import, ensure that `ENABLE_SENTRY` is set to false, and then show me the error message?

**Comment by vsoch** [comment]
@shea256 I've been debugging this for you, and I think I just found what might have been your issue. The RQScheduler has a sentry integration that is greedy to just use the token if it's discovered in settings, e.g., this line:

```
sentry_sdk.init(sentry_dsn, integrations=[RqIntegration()])
```
or [here])(https://github.com/rq/rq/blob/master/rq/contrib/sentry.py) and that needs to be disabled.

So - we basically need to figure out that 1) if the update does resolve the issue, why that is, or 2) if there is a programmatic way to disable the integration all together, given that the DSN isn't defined to be real.

**Comment by vsoch** [comment]
I think just putting the DSN within the boolean statement might work, testing it now.

**Comment by vsoch** [comment]
See https://github.com/vsoch/freegenes/commit/99298c4c32a4d20350bfcbcfb3f03f683256af8a#diff-d9b9b84282e9d139857d0cecfdbefc76

**Comment by shea256** [comment]
Hey @vsoch I just tried that and it does indeed work as expected.

However, shouldn't the DSN be defined outside of this file, in either config.py or secrets.py?

This problem should be solved as long as ENABLE_SENTRY is set to false, but a problem should arise if SENTRY_DSN isn't redefined in logging.py.

If you expect developers to only set SENTRY_DSN in logging.py then that is fine, but then we should make that clear in the documentation.

Personally I think it's easier to have just a single file for secrets and non-secret variables, and harder to remember / think to put just 1 variable in another file, but I'll defer to you here.

**Comment by vsoch** [comment]
When django initiates, it usually looks for a file called settings.py (take a look in different Django projects and you'll see it!) but instead, here we have a settings folder, and so the import first goes to settings/__init__.py. That file imports all of the files in the subfolder - they only exist actually to better organize the many configuration stuffs that a project needs. For example, we can isolate the main settings that the user cares about (and might need to change) in config.py. 

Generally:
 - All secrets go in secrets.py
 - All config settings (for a user) go in config.py
 - More detailed customizations can be done for any file, and you are right, it would be good to have this in the docs. And in fact it's there :) :point_right: https://vsoch.github.io/freegenes/docs/development/start#enabling-monitoring

**Comment by shea256** [comment]
> All secrets go in secrets.py
> All config settings (for a user) go in config.py

This makes sense, but then wouldn't we consider the DSN to be a config setting, and so it should be set in config.py?

**Comment by vsoch** [comment]
We would have to move the entire logging configuration over there, and I decided against this. The typical user not needing or wanting sentry it will work fine. The developer user is going to be going from top to bottom (we hope) through that docs page, and will find the note.

**Comment by shea256** [comment]
> We would have to move the entire logging configuration over there, and I decided against this.

Well not if it is defined in config.py and then imported into logging.py, correct?

**Comment by vsoch** [comment]
@shea256 I appreciate your feedback, the various files need to be cleanly imported into the single __init__.py. If you'd like to move forward, I'd be happy to finish up this review to integrate your changes.

**Comment by shea256** [comment]
@vsoch Sure thing, will move forward with the update to the PR.

**Comment by shea256** [comment]
@vsoch OK please check again, the code relating to sentry is gone now so the two remaining comments should be addressed.

**Comment by vsoch** [comment]
Great work @shea256 ! We're now ready for merge.

**Comment by vsoch** [comment]
Looks great! https://freegenes.dev/map/orders

**Comment by shea256** [comment]
Yes excellent, thank you!
